gpytorch/distributions/multivariate_normal.py

```diff
@@ -0,0 +1,169 @@
+        self._islazy = any(isinstance(arg, LazyTensor) for arg in (mean, covariance_matrix))
+        if self._islazy:
+            if validate_args:
+                # TODO: add argument validation
+                raise NotImplementedError()
+            self.loc = mean
+            self._covar = covariance_matrix
+            self.__unbroadcasted_scale_tril = None
+            self._validate_args = validate_args
+            batch_shape, event_shape = self.loc.shape[:-1], self.loc.shape[-1:]
+            # TODO: Integrate argument validation for LazyTensors into torch.distribution validation logic
+            super(TMultivariateNormal, self).__init__(batch_shape, event_shape, validate_args=False)
+        else:
+            super(MultivariateNormal, self).__init__(
+                loc=mean, covariance_matrix=covariance_matrix, validate_args=validate_args
+            )
+
+    @property
+    def _unbroadcasted_scale_tril(self):
+        if self.islazy and self.__unbroadcasted_scale_tril is None:
+            # cache root decoposition
+            self.__unbroadcasted_scale_tril = self.lazy_covariance_matrix.root_decomposition()
+        return self.__unbroadcasted_scale_tril
+
+    @_unbroadcasted_scale_tril.setter
+    def _unbroadcasted_scale_tril(self, ust):
+        if self.islazy:
+            raise NotImplementedError("Cannot set _unbroadcasted_scale_tril for lazy MVN distributions")
+        else:
+            self.__unbroadcasted_scale_tril = ust
+
+    def confidence_region(self):
+        """
+        Returns 2 standard deviations above and below the mean.
+
+        Returns:
+            Tuple[Tensor, Tensor]: pair of tensors of size (b x d) or (d), where
+                b is the batch size and d is the dimensionality of the random
+                variable. The first (second) Tensor is the lower (upper) end of
+                the confidence region.
+
+        """
+        std2 = self.stddev.mul_(2)
+        mean = self.mean
+        return mean.sub(std2), mean.add(std2)
+
+    @lazy_property
+    def covariance_matrix(self):
+        if self.islazy:
+            return self._covar.evaluate()
+        else:
+            return super(MultivariateNormal, self).covariance_matrix
+
+    def get_base_samples(self, sample_shape=torch.Size()):
+        """Get i.i.d. standard Normal samples (to be used with rsample(base_samples=base_samples))"""
+        with torch.no_grad():
+            shape = self._extended_shape(sample_shape)
+            base_samples = _standard_normal(shape, dtype=self.loc.dtype, device=self.loc.device)
+        return base_samples
+
+    @lazy_property
+    def lazy_covariance_matrix(self):
+        """
+        The covariance_matrix, represented as a LazyTensor
+        """
+        if self.islazy:
+            return self._covar
+        else:
+            return NonLazyTensor(super(MultivariateNormal, self).covariance_matrix)
+
+    def rsample(self, sample_shape=torch.Size(), base_samples=None):
```

Comment:
@gpleiss why are you re-implementing the sampling using the lazy variable? In my PR I used `rsample` from the torch `MultivariateNormal` that should handle re-shaping appropriately - that's the whole reason to give this here the `_unbroadcasted_scale_tril` method. Since it also calls `root_decomposition` under the hood just as `zero_mean_mvn_samples` it should be enough to just wrap the superclass method into one that deals with the base samples. Of course we should verify that any additional overhead incurred by using the torch Distribution objects doesn't slow things down much.

Reply:
Okay - I remember the reason that we added `zero_mean_mvn_samples` - for DiagLazyTensor and BlockDiagLazyTensor, `root_decomposition` is actually rather expensive and inaccurate, but in both cases there's a faster way to compute `zero_mean_mvn_samples`. (E.g. BlockDiagLazyTensor - we simply root-decompose the blocks).

Reply:
We use `BlockDiagLazyTensor` for additive kernels and SKI, so it's probably best to call `zero_mean_mvn_samples` in `root_decomposition`.

Reply:
The issue is with the low-rank approximation made by RootDecomposition. RootDecomposition iteratively forms a decomposition using Lanczos, where each column is added one at a time. Block diagonal matrices are naturally high rank (at least the rank is the number of blocks), so if we call root_decomposition directly on one - it takes many Lanczos iterations to get an accurate decomposition.

If we just care about computing samples, it's fastest to draw samples from each block independently (call root_decomposition on each of the blocks and multiply by independent noise vectors). This is because each block has a relatively low rank - requiring few Lanczos iterations - and all blocks can be decomposed in parallel. We discovered this when working on the LOVE paper.

I do think we don't want to override root_decomposition for BlockDiagonalLazyTensors however. The other place where we use root_decomposition is for product kernel decompositions, where we use it to create a low-rank approximation of the kernel. If I remember correctly, I think this is why we created separate methods for root_decomposition and zero_mean_mvn_samples. This might be worth revisiting though.